src/sage/knots/free_knotinfo_monoid.py

```diff
@@ -0,0 +1,479 @@
+        P = self.parent()
+        if len(wl) == 1:
+            name = wl[0]
+            L = P._index_dict[name][0].link()
+            if name.endswith(SymmetryMutant.mirror_image.value):
+                return L.mirror_image()
+            if name.endswith(SymmetryMutant.reverse.value):
+                return L.reverse()
+            if name.endswith(SymmetryMutant.concordance_inverse.value):
+                return L.mirror_image().reverse()
+            return L
+        else:
+            from sage.misc.misc_c import prod
+            return prod(P.gen(wl[i]).as_knot() for i in range(len(wl)))
+
+    def to_knotinfo(self):
+        r"""
+        Return a word representing ``self`` as a list of pairs where each pair
+        ``(ki, sym)`` consists of a :class:`~sage.knots.knotinfo.KontInfoBase` instance ``ki`` and
+        :class:`~sage.knots.knotinfo.SymmetryMutant` instance ``sym``.
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: FKIM.inject_variables(select=3)
+            Defining K3_1
+            Defining K3_1m
+            sage: K = K3_1^2 * K3_1m
+            sage: K.to_knotinfo()
+            [(<KnotInfo.K3_1: '3_1'>, <SymmetryMutant.itself: 's'>),
+            (<KnotInfo.K3_1: '3_1'>, <SymmetryMutant.itself: 's'>),
+            (<KnotInfo.K3_1: '3_1'>, <SymmetryMutant.mirror_image: 'm'>)]
+        """
+        wl = self.to_word_list()
+        P = self.parent()
+        return [P._index_dict[w] for w in wl]
+
+
+class FreeKnotInfoMonoid(IndexedFreeAbelianMonoid):
+
+    Element = FreeKnotInfoMonoidElement
+
+    @staticmethod
+    def __classcall_private__(cls, max_crossing_number=6, prefix=None, **kwds):
+        r"""
+        Normalize input to ensure a unique representation.
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FreeKnotInfoMonoid()
+            Free abelian monoid of knots with at most 6 crossings
+            sage: FreeKnotInfoMonoid(5)
+            Free abelian monoid of knots with at most 5 crossings
+        """
+        return super().__classcall__(cls, max_crossing_number, prefix=prefix, **kwds)
+
+    def __init__(self, max_crossing_number, category=None, prefix=None, **kwds):
+        r"""
+        Init this monoid with generators belonging to prime knots with at most
+        ``max_crossing_number`` crossings.
+
+        TESTS:
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: FKIM4 =  FreeKnotInfoMonoid(4)
+            sage: TestSuite(FKIM).run()
+            sage: TestSuite(FKIM4).run()
+        """
+        self._max_crossing_number = None
+        self._set_index_dictionary(max_crossing_number=max_crossing_number)
+        from sage.sets.set import Set
+        indices = Set(self._index_dict.keys())
```

Comment:
Not really important but does it need to be a Sage `Set`? I have an instinctual desire to avoid using them in general.

Reply:
Yes, because `IndexedMonoid` uses `cardinality` in `__init__`.

Reply:
Ah, I see. Then it is okay (an alternative might be `FiniteEnumeratedSet`). Do you also need the `.keys()` here too?

Reply:
> an alternative might be FiniteEnumeratedSet

Yes, that works. Done in the current commit.

> Do you also need the .keys() here too?

Yes for easier translation, for example in the `to_knotinfo` method.

Reply:
I don't quite follow with the `keys()`. A `dict` iterates over its keys and it is common Python practice to not have such calls. Well, if you prefer it, then we can leave it in, but don't be surprised if someone else comes and removes it in the future. Once you decide what you want to do (remove it or keep it), you can set a positive review.

Reply:
Sorry, I completely misunderstood that. Of course I can remove the `.keys()` here.

Reply:
Thank you.

Reply:
Thank you, too.